lib/Dialects/LinalgExt/Transforms/InParallelToSequentialFor.cpp

```diff
@@ -0,0 +1,115 @@
+//===- InParallelToSequentialFor.cpp.cpp - Rewrite InParallel as ForOp ---===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===---------------------------------------------------------------------===//
+
+#include <mlir/IR/BlockAndValueMapping.h>
+#include <mlir/IR/BuiltinOps.h>
+
+#include "Dialects/LinalgExt/LinalgExtOps.h"
+#include "Dialects/LinalgExt/PassDetail.h"
+#include "Dialects/LinalgExt/Passes.h"
+#include "Transforms/Utils.h"
+#include "llvm/ADT/STLExtras.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Identifier.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::linalg_ext;
+
+namespace {
+
+struct InParallelOpToSCFRewriter
+    : public OpRewritePattern<linalg_ext::InParallelOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg_ext::InParallelOp inParallelOp,
+                                PatternRewriter &rewriter) const override {
+    // Construct the loop bounds based on the canonical arithmetic progression.
+    Location loc = inParallelOp.getLoc();
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value numThreads = inParallelOp.num_threads();
+
+    // Construct the op without a body builder: we need to clone the ops in the
+    // body explicitly after having access to the new bbArgs.
+    // As a consequence, `ensureTerminator` is not called and the body has no
+    // terminator.
+    PerformConcurrentlyOp performConcurrentlyOp = inParallelOp.getTerminator();
+    scf::ForOp forOp = rewriter.create<scf::ForOp>(
+        loc, zero, numThreads, one, performConcurrentlyOp.yieldedValues());
+
+    rewriter.setInsertionPointToStart(forOp.getBody());
+
+    BlockAndValueMapping bvm;
+    bvm.map(performConcurrentlyOp.yieldedValues(), forOp.getRegionIterArgs());
```

Comment:
Perhaps move this after `mergeBlock`? It's placement is a little confusing, because it seems like this mapping might influence block merging but it really shouldn't.